File: anvil/src/test/java/trikita/anvil/XmlTest.java
package trikita.anvil;

import android.content.Context;
import android.view.View;

import org.junit.Test;

import static junit.framework.Assert.assertEquals;
import static trikita.anvil.BaseDSL.attr;
import static trikita.anvil.BaseDSL.xml;
import static trikita.anvil.BaseDSL.v;

public class XmlTest extends Utils {
    private final static int ID_HEADER = 100;
    private final static int LAYOUT = 1;

    private int inflateCount = 0;

    public XmlTest() {
        Anvil.registerViewFactory(new Anvil.ViewFactory() {
            public View fromClass(Context c, Class<? extends View> v) { return null; }
            public View fromXml(Context c, int xmlId) {
                System.out.println("fromXml " + c + " " + xmlId);
                if (xmlId == LAYOUT) {
                    inflateCount++;
                    MockLayout layout = new MockLayout(c);
                    MockView header = new MockView(c);
                    header.setId(ID_HEADER);
                    layout.addView(header, 0);
                    return layout;
                }
                return null;
            }
        });
    }

    @Test
    public void testXml() {
        Anvil.mount(container, new Anvil.Renderable() {
            public void view() {
                xml(LAYOUT, new Anvil.Renderable() {
                    public void view() {
                        attr("text", "foo");
                        v(MockView.class, new Anvil.Renderable() {
                            public void view() {
                                attr("tag", "bar");
                            }
                        });
                    }
                });
            }
        });
        assertEquals(1, inflateCount);
        MockLayout layout = (MockLayout) container.getChildAt(0);
        assertEquals("foo", layout.getText());
        assertEquals(1, layout.getChildCount());
        assertEquals(ID_HEADER, layout.getChildAt(0).getId());
        assertEquals("bar", layout.getChildAt(0).getTag());
    }
}

File: anvil/src/test/kotlin/dev/inkremental/XmlTest.kt
package dev.inkremental

import android.content.Context
import android.view.View
import android.view.ViewGroup
import kotlin.test.Test
import kotlin.test.assertEquals

class XmlTest : Utils() {
    private var inflateCount = 0
    @Test
    fun testXml() {
        Inkremental.mount(container, Renderable {
            xml(LAYOUT) {
                attr("text", "foo")
                v<MockView> { attr("tag", "bar") }
            }
        })
        assertEquals(1, inflateCount)
        val layout = container.getChildAt(0) as MockLayout
        assertEquals("foo", layout.text)
        assertEquals(1, layout.childCount)
        assertEquals(ID_HEADER, layout.getChildAt(0).id)
        assertEquals("bar", layout.getChildAt(0).tag)
    }

    companion object {
        private const val ID_HEADER = 100
        private const val LAYOUT = 1
    }

    init {
        Inkremental.registerViewFactory(object : Inkremental.ViewFactory {
            override fun fromClass(c: Context?, v: Class<out View?>): View? = null

            override fun fromXml(parent: ViewGroup, xmlId: Int): View? {
                if (xmlId == LAYOUT) {
                    inflateCount++
                    val layout = MockLayout(parent.context)
                    val header = MockView(parent.context)
                    header.id = ID_HEADER
                    layout.addView(header, 0)
                    return layout
                }
                return null
            }
        })
    }
}
